client/src/main/scala/org/http4s/client/websocket/WSClient.scala

```diff
@@ -0,0 +1,179 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.client.websocket
+
+import cats._
+import cats.data.{Chain, OptionT}
+import cats.effect._
+import cats.effect.concurrent.{Deferred, Ref, TryableDeferred}
+import cats.implicits._
+import fs2.{Pipe, Stream}
+import org.http4s.{Headers, Method, Uri}
+import scodec.bits.ByteVector
+
+/** A websocket request.
+  *
+  * @param uri
+  *   The URI.
+  * @param headers
+  *   The headers to send. Put your `Sec-Websocket-Protocol` headers here if needed. Some websocket
+  *   clients reject other WS-specific headers.
+  * @param method
+  *   The method of the intial HTTP request. Ignored by some clients.
+  */
+case class WSRequest(
+    uri: Uri,
+    headers: Headers = Headers.empty,
+    method: Method = Method.GET
+)
+
+sealed trait WSFrame extends Product with Serializable
+
+sealed trait WSControlFrame extends WSFrame
+
+sealed trait WSDataFrame extends WSFrame
+
+object WSFrame {
+  final case class Close(statusCode: Int, reason: String) extends WSControlFrame
+  final case class Ping(data: ByteVector) extends WSControlFrame
+  final case class Pong(data: ByteVector) extends WSControlFrame
+  final case class Text(data: String, last: Boolean = true) extends WSDataFrame
+  final case class Binary(data: ByteVector, last: Boolean = true) extends WSDataFrame
+}
+
+trait WSConnection[F[_]] {
+
+  /** Send a single websocket frame. The sending side of this connection has to be open. */
+  def send(wsf: WSFrame): F[Unit]
+
+  /** Send multiple websocket frames. Equivalent to multiple `send` calls, but at least as fast. */
+  def sendMany[G[_]: Foldable, A <: WSFrame](wsfs: G[A]): F[Unit]
+
+  /** A `Pipe` which sends websocket frames and emits a `()` for each chunk sent. */
+  final def sendPipe: Pipe[F, WSFrame, Unit] = _.chunks.evalMap(sendMany(_))
+
+  /** Wait for a single websocket frame to be received. Returns `None` if the receiving side is
+    * closed.
+    */
+  def receive: F[Option[WSFrame]]
+
+  /** A stream of the incoming websocket frames. */
+  final def receiveStream: Stream[F, WSFrame] = Stream.repeatEval(receive).unNoneTerminate
+
+  /** The negotiated subprotocol, if any. */
+  def subprotocol: Option[String]
+}
+
+trait WSConnectionHighLevel[F[_]] {
+
+  /** Send a single websocket frame. The sending side of this connection has to be open. */
+  def send(wsf: WSDataFrame): F[Unit]
+
+  /** Send multiple websocket frames. Equivalent to multiple `send` calls, but at least as fast. */
+  def sendMany[G[_]: Foldable, A <: WSDataFrame](wsfs: G[A]): F[Unit]
+
+  /** A `Pipe` which sends websocket frames and emits a `()` for each chunk sent. */
+  final def sendPipe: Pipe[F, WSDataFrame, Unit] = _.chunks.evalMap(sendMany(_))
+
+  /** Send a Ping frame. */
+  def sendPing(data: ByteVector = ByteVector.empty): F[Unit]
```

Comment:
IDK if it's too restrictive, but if possible I'd like the high-level API to be fully implementable with browser websockets:
https://developer.mozilla.org/en-US/docs/Web/API/WebSocket

I think this `sendPing` is the only method in the current interface that can't be implemented in browsers. Is it necessary? :)

Reply:
Thanks @armanbilge for getting the ball rolling on this.

I agree that we should have an interface that's fully implementable with browser websockets.

Maybe we should split the high level and low level interfaces?

Reply:
Yes, so the current `WSClient` looks like this:
```scala
trait WSClient[F[_]] {
  def connect(request: WSRequest): Resource[F, WSConnection[F]]
  def connectHighLevel(request: WSRequest): Resource[F, WSConnectionHighLevel[F]]
}
```

The `WSConnectionHighLevel` I think is fully-implementable with browser websockets, except for that `sendPing` method I noted above.

So the question is what to do about the non-highlevel `connect` method. We could simply throw an error if tried to use in the browser. Or for additional type-safety we could split:
```scala
trait WSHighLevelClient[F[_]] {
  def connectHighLevel(request: WSRequest): Resource[F, WSConnectionHighLevel[F]]
}

trait WSClient[F[_]] extends WSHighLevelClient[F] {
  def connect(request: WSRequest): Resource[F, WSConnection[F]]
}
```

I'm not sure how these things get used by clients, but I'd even think maybe the distinction should be `WSClient` and `WSLowLevelClient` if most people prefer the high-level interface?

Reply:
Does the browser ping by itself. How do people prevent themselves being killed as an inactive tcp session? You can have a ping data packet, but its... repetitive.

Reply:
That's a good question, I don't know. Probably? Can't find any information about that.

Reply:
https://www.w3.org/Bugs/Public/show_bug.cgi?id=13104

> We don't think it's necessary to include API support for ping/pong, which is similar to TCP-level-like keepalives that can cross intermediaries. They are used to notify anyone on the path to the server and the server itself that the client is still alive.
 
> It should be the job of the stack to keep your connection alive. The application may want to specify keepalive interval, but it should not need to handle this task itself.